packages/dd-trace/src/profiling/profilers/wall.js

```diff
@@ -1,23 +1,111 @@
 'use strict'
 
+const { storage } = require('../../../../datadog-core')
+
+const dc = require('../../../../diagnostics_channel')
+
+const beforeCh = dc.channel('dd-trace:storage:before')
+const enterCh = dc.channel('dd-trace:storage:enter')
+
+let kSampleCount
+
+function getActiveSpan () {
+  const store = storage.getStore()
+  return store && store.span
+}
+
+function getStartedSpans (context) {
+  return context._trace.started
+}
+
+function generateLabels ({ spanId, rootSpanId, webTags, endpoint }) {
+  const labels = {}
+  if (spanId) {
+    labels['span id'] = spanId
+  }
+  if (rootSpanId) {
+    labels['local root span id'] = rootSpanId
+  }
+  if (webTags && Object.keys(webTags).length !== 0) {
+    labels['trace endpoint'] = endpointNameFromTags(webTags)
+  } else if (endpoint) {
+    // fallback to endpoint computed when sample was taken
+    labels['trace endpoint'] = endpoint
+  }
+
+  return labels
+}
+
+function getSpanContextTags (span) {
+  return span.context()._tags
+}
+
+function isWebServerSpan (tags) {
+  return tags['span.type'] === 'web'
+}
+
+function endpointNameFromTags (tags) {
+  return tags['resource.name'] || [
+    tags['http.method'],
+    tags['http.route']
+  ].filter(v => v).join(' ')
+}
+
+function updateContext (context, span, startedSpans, endpointCollectionEnabled) {
+  context.spanId = span.context().toSpanId()
+  const rootSpan = startedSpans[0]
+  if (rootSpan) {
+    context.rootSpanId = rootSpan.context().toSpanId()
+    if (endpointCollectionEnabled) {
+      for (let i = startedSpans.length - 1; i >= 0; i--) {
```

Comment:
Old code (in since deleted `cpu.js`) was walking the array from the front, I think, essentially trying to find the span closest to root that is a web span. Not sure if it makes a difference what direction is the array traversed - not sure if there's more than one such span in it.

Reply:
Thanks for the explanation!

Reply:
I added a comment in the code for future me ;)